mempool: trap on double-free instances

npitre opened the pull request:
A double-free could cause very hard to find bugs when using the mempool
allocator as the same memory would end up being allocated twice
afterwards.

Now that bits in the block bitmap are cleared only when actually freeing
a block, we may simply ensure those bits are still set before clearing
them, effectively catching most double-free cases.

The alloc_bit_is_set() function is made static inline so that when
assertion checks are disabled the compiler won't complain about unused
code.

npitre committed:
mempool: trap on double-free instances

A double-free could cause very hard to find bugs when using the mempool
allocator as the same memory would end up being allocated twice
afterwards.

Now that bits in the block bitmap are cleared only when actually freeing
a block, we may simply ensure those bits are still set before clearing
them, effectively catching most double-free cases.

The alloc_bit_is_set() function is made static inline so that when
assertion checks are disabled the compiler won't complain about unused
code.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

andrewboie commented:
Hmm, looks like this patch may have found some stuff based on the posted CI failure.
Run tests under tests/kernel/pipe/ to see what I mean

npitre commented:
On Fri, 4 Oct 2019, Andrew Boie wrote:

> Hmm, looks like this patch may have found some stuff based on the posted CI failure.
> Run tests under tests/kernel/pipe/ to see what I mean

I filed #19535 already, hoping that someone more knowledgeable with 
pipes would check it out.

In the mean time, this CI failure shouldn't prevent the merging of 
this PR. This is not creating a bug but rather exposing an existing one.

andrewboie commented:
> In the mean time, this CI failure shouldn't prevent the merging of
this PR. This is not creating a bug but rather exposing an existing one

Unfortunately, unless the test is also temporarily disabled, simply merging this will cause CI to be in a non-passing state, which we can't have.

However, in this case the fix is easy; the extra free in the test needs to be removed.

andrewboie commented:
Hmm, after looking at the pipe code some more, I am less sure. But at any rate we need to either disable the test or fix the issue

pabigot commented on lib/os/mempool.c:
```diff
@@ -56,6 +56,14 @@ static void clear_alloc_bit(struct sys_mem_pool_base *p, int level, int bn)
+	u32_t *word;
+	int bit = get_bit_ptr(p, level, bn, &word);
+
+	return !!(*word & (1<<bit));
```
`(1 << bit)` should be spelled `BIT(bit)` using `sys/util.h` which correctly handles the case of the high bit of an unsigned type.  (`1 << 31` is undefined for signed 32-bit systems.)

It appears there are many other occurrences of `(1 << bit)` in this file that should also be fixed.

If MISRA compliance is important the return value should be something like `(*word & BIT(bit)) != 0U)`.

npitre commented:
The double-free issue in CI is now fixed, and I just pushed a slightly 
more efficient version that avoids the (1 << bit) construct. That won't 
make MISRA happy about the rest of the file, but that really ought to be 
addressed in a separate PR.
